src/util/ProjectIO.as

```diff
@@ -207,48 +207,113 @@ public class ProjectIO {
 			}
-			for each (c in allCostumes) c.generateOrFindComposite(allCostumes);
-			whenDone();
 		}
+		// decrement the artificial 1 and also handle the case where no images are present.
+		imageDone();
+	}
 
-		var c:ScratchCostume;
-		var allCostumes:Array = [];
-		for each (var o:ScratchObj in objList) {
-			for each (c in o.costumes) allCostumes.push(c);
+	private function allImagesLoaded(objList:Array, imageDict:Dictionary, whenDone:Function, fail:Function):void {
+		var errorCostumes:Vector.<ScratchCostume> = new Vector.<ScratchCostume>(2);
+
+		function makeErrorImage(obj:ScratchObj, c:ScratchCostume):* {
+			// TODO: translation?
+			if (!errorDialog) {
+				errorDialog = new DialogBox();
+				errorDialog.addTitle('Image loading error');
+				errorDialog.addText('At least one backdrop or costume failed to load:\n');
+			}
+			if (obj.isStage) {
+				errorDialog.addText('Backdrop: ' + c.costumeName + '\n');
+			}
+			else {
+				errorDialog.addText('Sprite: ' + obj.objName + '\nCostume: ' + c.costumeName + '\n');
+			}
+
+			var errorCostumeIndex:int = 0 + obj.isStage;
+			var errorCostume:ScratchCostume = errorCostumes[errorCostumeIndex] =
+					errorCostumes[errorCostumeIndex] || ScratchCostume.emptyBitmapCostume('', obj.isStage);
+			return errorCostume.baseLayerBitmap;
+		}
+
+		var allCostumes:Vector.<ScratchCostume> = new <ScratchCostume>[];
+		var errorDialog:DialogBox;
+		var img:*; // either BitmapData or SVGElement
+
+		for each (var obj:ScratchObj in objList) {
+			for each (var c:ScratchCostume in obj.costumes) {
+				allCostumes.push(c);
+				if ((c.baseLayerData != null) && (c.baseLayerBitmap == null)) {
+					img = imageDict[c.baseLayerData];
+					if (!img) {
+						c.baseLayerBitmap = makeErrorImage(obj, c);
+						c.baseLayerData = null;
+					}
+					else if (img is BitmapData) {
+						c.baseLayerBitmap = img;
+					}
+					else if (img is SVGElement) {
+						c.setSVGRoot(img, false);
+					}
+				}
+				if ((c.textLayerData != null) && (c.textLayerBitmap == null)) {
+					img = imageDict[c.textLayerData];
+					if (img) {
+						c.textLayerBitmap = imageDict[c.textLayerData];
+					}
+					else {
+						c.textLayerBitmap = makeErrorImage(obj, c);
+						c.textLayerData = null;
+					}
+				}
+			}
 		}
-		var imageDict:Dictionary = new Dictionary(); // maps image data to BitmapData
 		for each (c in allCostumes) {
-			if ((c.baseLayerData != null) && (c.baseLayerBitmap == null)) {
-				if (ScratchCostume.isSVGData(c.baseLayerData)) decodeSVG(c.baseLayerData, imageDict, imageDecoded);
-				else decodeImage(c.baseLayerData, imageDict, imageDecoded, decodeError);
+			c.generateOrFindComposite(allCostumes);
+		}
+
+		if (errorDialog) {
+			errorDialog.addButton('OK', errorDialog.accept);
+			errorDialog.showOnStage(Scratch.app.stage);
+
+			if (fail != null) {
+				fail();
+			}
+			else if (whenDone != null) {
+				whenDone();
```

Comment:
Some callers provide a failure callback, and some expect `whenDone()` to be called even on failure. Failing to call `whenDone()` in the failure case was basically the core cause of #1085.
While it would be nice to make callers consistent, I chose to do this solution instead of making this PR even larger...